src/vm/gchelpers.cpp

```diff
@@ -408,15 +420,127 @@ inline SIZE_T MaxArrayLength(SIZE_T componentSize)
     return (componentSize == 1) ? 0X7FFFFFC7 : 0X7FEFFFFF;
 }
 
-OBJECTREF AllocateValueSzArray(TypeHandle elementType, INT32 length)
+OBJECTREF AllocateSzArray(TypeHandle arrayType, INT32 cElements, BOOL zeroingOptional)
 {
     CONTRACTL {
         THROWS;
         GC_TRIGGERS;
         MODE_COOPERATIVE; // returns an objref without pinning it => cooperative        
     } CONTRACTL_END;
 
-    return AllocateArrayEx(elementType.MakeSZArray(), &length, 1);
+    SetTypeHandleOnThreadForAlloc(arrayType);
```

Comment:
All other places that call `SetTypeHandleOnThreadForAlloc` seems to use the MethodTable. Why to make this one different?

Reply:
They do `SetTypeHandleOnThreadForAlloc(TypeHandle(pArrayMT))`  and the pArrayMT is passed from the caller which fetches it from the typehandle.

Here I would end up with:

```cs
    ArrayTypeDesc* arrayDesc = arrayType.AsArray();
    MethodTable* pArrayMT = arrayDesc->GetMethodTable();
    SetTypeHandleOnThreadForAlloc(TypeHandle(pArrayMT))
```

Is that potentially different from the following? 

```cs
   SetTypeHandleOnThreadForAlloc(arrayType)
```



---
In reply to: [279240429](https://github.com/dotnet/coreclr/pull/24096#discussion_r279240429) [](ancestors = 279240429)

Reply:
Yes, it is different. We have both MethodTable* and TypeDesc for arrays. The first one will set the "TypeHandleOnThreadForAlloc" to MethodTable*, the second one sets it to TypeDesc.

Reply:
It is a bit confusing. I kind of expected that TypeHandle(pArrayMT) converts it back to TypeDesc form. Will change.

---
In reply to: [279464064](https://github.com/dotnet/coreclr/pull/24096#discussion_r279464064) [](ancestors = 279464064)